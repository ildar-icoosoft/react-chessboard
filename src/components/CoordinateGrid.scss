.coordinateGrid {
  cursor: pointer;
  background-image: url("../images/brown.svg");
  position: absolute;
  background-size: cover;
}
.svg {
  overflow: hidden;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  pointer-events: none;
  z-index: 2;
  opacity: 0.6;
}
