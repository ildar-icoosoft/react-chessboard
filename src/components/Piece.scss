.piece {
  position: absolute;
  top: 0;
  left: 0;
  width: 12.5%;
  height: 12.5%;
  background-size: cover;
  z-index: 2;
  will-change: transform;
  pointer-events: none;
}

.bB {
  background-image: url(../images/pieces/black_bishop.svg);
}
.bK {
  background-image: url(../images/pieces/black_king.svg);
}
.bN {
  background-image: url(../images/pieces/black_knight.svg);
}
.bP {
  background-image: url(../images/pieces/black_pawn.svg);
}
.bQ {
  background-image: url(../images/pieces/black_queen.svg);
}
.bR {
  background-image: url(../images/pieces/black_rook.svg);
}

.wB {
  background-image: url(../images/pieces/white_bishop.svg);
}
.wK {
  background-image: url(../images/pieces/white_king.svg);
}
.wN {
  background-image: url(../images/pieces/white_knight.svg);
}
.wP {
  background-image: url(../images/pieces/white_pawn.svg);
}
.wQ {
  background-image: url(../images/pieces/white_queen.svg);
}
.wR {
  background-image: url(../images/pieces/white_rook.svg);
}
