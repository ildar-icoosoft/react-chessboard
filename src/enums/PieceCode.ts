export enum PieceCode {
  WHITE_PAWN = "wP",
  BLACK_PAWN = "bP",
  WHITE_KNIGHT = "wN",
  BLACK_KNIGHT = "bN",
  WHITE_BISHOP = "wB",
  BLACK_BISHOP = "bB",
  WHITE_ROOK = "wR",
  BLACK_ROOK = "bR",
  WHITE_QUEEN = "wQ",
  BLACK_QUEEN = "bQ",
  WHITE_KING = "wK",
  BLACK_KING = "bK",
}
