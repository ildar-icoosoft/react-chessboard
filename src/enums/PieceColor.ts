export enum PieceColor {
  WHITE = "white",
  BLACK = "black",
}
