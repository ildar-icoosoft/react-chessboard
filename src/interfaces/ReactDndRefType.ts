import { DragDropManager } from "dnd-core";

export interface ReactDndRefType {
  getManager(): DragDropManager | undefined;
  getDecoratedComponent<T>(): T;
}
