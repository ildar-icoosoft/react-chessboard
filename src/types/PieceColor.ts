export type PieceColor = "white" | "black";
