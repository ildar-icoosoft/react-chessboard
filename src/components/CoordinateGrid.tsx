import React, {
  forwardRef,
  MouseEvent,
  RefObject,
  useImperativeHandle,
  useRef,
} from "react";
import { PieceColor } from "../enums/PieceColor";
import { Position } from "../interfaces/Position";
import { toPairs as _toPairs } from "lodash";
import { Piece } from "./Piece";
import css from "./CoordinateGrid.scss";
import { DEFAULT_BOARD_WIDTH } from "../constants/constants";
import {
  getSquareAlgebraicCoordinates,
  getSquareXYCoordinates,
} from "../utils/chess";
import { useDrop } from "react-dnd";
import { DragItemType } from "../enums/DragItemType";
import { useCombinedRefs } from "../hooks/useCombinedRefs";
import { Identifier } from "dnd-core";

export interface CoordinateGridRef {
  getDropHandlerId(): Identifier | null;
}

export interface CoordinateGridProps {
  orientation?: PieceColor;
  position?: Position;
  width?: number;

  onClick?(coordinates: string): void;
  onRightClick?(coordinates: string): void;
}

export const CoordinateGrid = forwardRef<
  CoordinateGridRef,
  CoordinateGridProps
>(
  (
    {
      position = {},
      width = DEFAULT_BOARD_WIDTH,
      orientation = PieceColor.WHITE,
      onClick,
      onRightClick,
    },
    ref
  ) => {
    const domRef: RefObject<HTMLDivElement> = useRef<HTMLDivElement>(null);

    const handleClick = (event: MouseEvent): void => {
      if (onClick) {
        const rect: DOMRect = (domRef.current as HTMLDivElement).getBoundingClientRect();

        const coordinates: string = getSquareAlgebraicCoordinates(
          {
            x: event.clientX - rect.left,
            y: event.clientY - rect.top,
          },
          width,
          orientation
        );

        onClick(coordinates);
      }
    };

    const handleContextMenu = (event: MouseEvent): void => {
      event.preventDefault();

      if (onRightClick) {
        const rect: DOMRect = (domRef.current as HTMLDivElement).getBoundingClientRect();

        const coordinates: string = getSquareAlgebraicCoordinates(
          {
            x: event.clientX - rect.left,
            y: event.clientY - rect.top,
          },
          width,
          orientation
        );

        onRightClick(coordinates);
      }
    };

    const [{ dropHandlerId }, dropRef] = useDrop({
      accept: DragItemType.PIECE,
      collect(monitor) {
        return {
          dropHandlerId: monitor.getHandlerId(),
        };
      },
    });

    useImperativeHandle(ref, () => ({
      getDropHandlerId: (): Identifier | null => dropHandlerId,
    }));

    return (
      <div
        data-testid={"coordinate-grid"}
        className={css.coordinateGrid}
        style={{ width: `${width}px`, height: `${width}px` }}
        onClick={handleClick}
        onContextMenu={handleContextMenu}
        ref={useCombinedRefs(dropRef, domRef)}
      >
        {_toPairs(position).map((pair) => (
          <Piece
            pieceCode={pair[1]}
            width={width / 8}
            xYCoordinates={getSquareXYCoordinates(pair[0], width, orientation)}
            key={pair[0]}
          />
        ))}
      </div>
    );
  }
);
