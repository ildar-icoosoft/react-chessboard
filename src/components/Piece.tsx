import React, { CSSProperties, FC } from "react";
import { PieceCode } from "../enums/PieceCode";
import { getPieceElement } from "../utils/chess";
import { Transition } from "react-transition-group";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";
import { TransitionStatus } from "react-transition-group/Transition";
import { PartialRecord } from "../types/PartialRecord";
import {
  DEFAULT_SQUARE_WIDTH,
  DEFAULT_TRANSITION_DURATION,
} from "../constants/constants";

export interface PieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionFrom?: SquareTransitionFrom;
  transitionDuration?: number;
}

const getTransitionStyles = (
  transitionFrom: SquareTransitionFrom | undefined,
  state: TransitionStatus,
  transitionDuration: number
): CSSProperties => {
  if (!transitionFrom) {
    return {};
  }

  const styles: PartialRecord<TransitionStatus, CSSProperties> = {
    entering: {
      transform: `translate(${transitionFrom.x}px, ${transitionFrom.y}px)`,
    },
    entered: {
      transform: "translate(0, 0)",
      transition: `transform ${transitionDuration}ms`,
      zIndex: 1,
    },
  };

  return styles[state] || {};
};

export const Piece: FC<PieceProps> = ({
  pieceCode,
  width = DEFAULT_SQUARE_WIDTH,
  transitionFrom,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
}) => {
  return (
    <Transition appear={true} in={transitionFrom ? true : false} timeout={0}>
      {(state) => {
        return (
          <div
            className={"piece"}
            data-testid={`piece-${pieceCode}`}
            style={getTransitionStyles(
              transitionFrom,
              state,
              transitionDuration
            )}
          >
            <svg viewBox={`1 1 43 43`} width={width} height={width}>
              <g>{getPieceElement(pieceCode, { width: 45, height: 45 })}</g>
            </svg>
          </div>
        );
      }}
    </Transition>
  );
};
