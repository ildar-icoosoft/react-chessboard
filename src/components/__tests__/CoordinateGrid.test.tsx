import React, { createRef } from "react";
import TestRenderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import { CoordinateGrid, CoordinateGridRef } from "../CoordinateGrid";
import { Piece } from "../Piece";
import { PieceCode } from "../../enums/PieceCode";
import { fireEvent, render, createEvent } from "@testing-library/react";
import { PieceColor } from "../../enums/PieceColor";
import { wrapInTestContext } from "react-dnd-test-utils";
import { ReactDndRefType } from "../../interfaces/ReactDndRefType";

jest.useFakeTimers();

describe("CoordinateGrid", () => {
  const CoordinateGridWithDnd = wrapInTestContext(CoordinateGrid);

  it("Snapshot", () => {
    const tree = TestRenderer.create(<CoordinateGridWithDnd />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains 1 Piece", () => {
      const testRenderer = TestRenderer.create(<CoordinateGridWithDnd />);
      const testInstance = testRenderer.root;

      expect(testInstance.findAllByType(Piece).length).toBe(0);

      testRenderer.update(
        <CoordinateGridWithDnd
          position={{ e2: PieceCode.WHITE_PAWN, d5: PieceCode.BLACK_BISHOP }}
        />
      );

      expect(
        testInstance.findAll((item) => {
          return (
            item.type === Piece && item.props.pieceCode === PieceCode.WHITE_PAWN
          );
        }).length
      ).toBe(1);

      expect(
        testInstance.findAll((item) => {
          return (
            item.type === Piece &&
            item.props.pieceCode === PieceCode.BLACK_BISHOP
          );
        }).length
      ).toBe(1);
    });
  });

  describe("children components props", () => {
    describe("Piece", () => {
      it("pieceCode", () => {
        const testInstance = TestRenderer.create(
          <CoordinateGridWithDnd position={{ e2: PieceCode.WHITE_PAWN }} />
        ).root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.pieceCode).toBe(PieceCode.WHITE_PAWN);
      });

      it("width", () => {
        const testRenderer = TestRenderer.create(
          <CoordinateGridWithDnd position={{ e2: PieceCode.WHITE_PAWN }} />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.width).toBe(60);

        testRenderer.update(
          <CoordinateGridWithDnd
            position={{ e2: PieceCode.WHITE_PAWN }}
            width={240}
          />
        );
        expect(piece.props.width).toBe(30);
      });

      it("xYCoordinates", () => {
        const testRenderer = TestRenderer.create(
          <CoordinateGridWithDnd position={{ e2: PieceCode.WHITE_PAWN }} />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.xYCoordinates).toEqual({
          x: 240,
          y: 360,
        });

        testRenderer.update(
          <CoordinateGridWithDnd
            position={{ e2: PieceCode.WHITE_PAWN }}
            width={240}
          />
        );
        expect(piece.props.xYCoordinates).toEqual({
          x: 120,
          y: 180,
        });
      });
    });
  });

  describe("Events", () => {
    it("Click", () => {
      const onClick = jest.fn();
      const onRightClick = jest.fn();

      const { getByTestId, rerender } = render(
        <CoordinateGridWithDnd onClick={onClick} onRightClick={onRightClick} />
      );

      const coordinateGridEl = getByTestId("coordinate-grid");

      fireEvent.click(coordinateGridEl, {
        clientX: 60,
        clientY: 60,
      });

      rerender(
        <CoordinateGridWithDnd
          onClick={onClick}
          onRightClick={onRightClick}
          orientation={PieceColor.BLACK}
        />
      );

      fireEvent.click(coordinateGridEl, {
        clientX: 479,
        clientY: 0,
      });

      expect(onClick).toHaveBeenCalledTimes(2);

      expect(onClick).toHaveBeenNthCalledWith(1, "b7");
      expect(onClick).toHaveBeenNthCalledWith(2, "a1");

      expect(onRightClick).toHaveBeenCalledTimes(0);
    });

    it("Click if no callback", () => {
      const { getByTestId } = render(<CoordinateGridWithDnd />);
      expect(() => {
        fireEvent.click(getByTestId("coordinate-grid"));
      }).not.toThrow();
    });

    it("Right Click", () => {
      const onClick = jest.fn();
      const onRightClick = jest.fn();

      const { getByTestId, rerender } = render(
        <CoordinateGridWithDnd onClick={onClick} onRightClick={onRightClick} />
      );

      const coordinateGridEl = getByTestId("coordinate-grid");

      fireEvent.contextMenu(coordinateGridEl, {
        clientX: 60,
        clientY: 60,
      });

      rerender(
        <CoordinateGridWithDnd
          onClick={onClick}
          onRightClick={onRightClick}
          orientation={PieceColor.BLACK}
        />
      );

      fireEvent.contextMenu(coordinateGridEl, {
        clientX: 479,
        clientY: 0,
      });

      expect(onRightClick).toHaveBeenCalledTimes(2);

      expect(onRightClick).toHaveBeenNthCalledWith(1, "b7");
      expect(onRightClick).toHaveBeenNthCalledWith(2, "a1");

      expect(onClick).toHaveBeenCalledTimes(0);
    });

    it("Right Click event must be prevented", () => {
      const { getByTestId } = render(<CoordinateGridWithDnd />);

      const coordinateGridEl = getByTestId("coordinate-grid");

      const contextMenuEvent = createEvent.contextMenu(coordinateGridEl, {
        clientX: 479,
        clientY: 0,
      });

      fireEvent(coordinateGridEl, contextMenuEvent);

      expect(contextMenuEvent.defaultPrevented).toBeTruthy();
    });

    it("Right Click if no callback", () => {
      const { getByTestId } = render(<CoordinateGridWithDnd />);
      expect(() => {
        fireEvent.contextMenu(getByTestId("coordinate-grid"));
      }).not.toThrow();
    });
  });

  describe("methods", () => {
    it("getDropHandlerId()", () => {
      const dragAndDropRef = createRef<ReactDndRefType>();

      TestRenderer.create(<CoordinateGridWithDnd ref={dragAndDropRef} />);

      const coordinateGridRef: CoordinateGridRef = (dragAndDropRef.current as ReactDndRefType).getDecoratedComponent<
        CoordinateGridRef
      >();

      expect(() => coordinateGridRef.getDropHandlerId()).toBeTruthy();
    });
  });

  describe("DOM structure", () => {
    it("should contain data-testid coordinateGrid", () => {
      const { queryByTestId } = render(<CoordinateGridWithDnd />);
      expect(queryByTestId("coordinate-grid")).toBeInTheDocument();
    });

    it("coordinateGrid should have width and height styles", () => {
      const { getByTestId, rerender } = render(<CoordinateGridWithDnd />);

      const coordinateGridEl = getByTestId("coordinate-grid");

      expect(coordinateGridEl).toHaveStyle({
        width: "480px",
        height: "480px",
      });

      rerender(<CoordinateGridWithDnd width={700} />);

      expect(coordinateGridEl).toHaveStyle({
        width: "700px",
        height: "700px",
      });
    });
  });
});
