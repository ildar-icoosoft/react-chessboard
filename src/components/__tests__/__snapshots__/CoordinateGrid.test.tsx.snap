// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoordinateGrid Snapshot 1`] = `
<div
  className="coordinateGrid"
  data-testid="coordinate-grid"
  onClick={[Function]}
  onContextMenu={[Function]}
  style={
    Object {
      "height": "480px",
      "width": "480px",
    }
  }
/>
`;
